src/visitable.h

```diff
@@ -60,6 +59,29 @@ class visitable
         /** Returns true if any item (including those within a container) matches the filter */
         bool has_item_with( const std::function<bool( const item & )> &filter ) const;
 
+        /** Returns true if instance has amount (or more) items of at least quality level */
+        bool has_items_with_quality( const std::string &qual, int level = 1, int amount = 1 ) const;
+
+        /** Count maximum available charges from this instance and any contained items */
+        long charges_of( const std::string &what ) const;
+
+        /**
+         * Count items matching id including both this instance and any contained items
+         * @param pseudo whether pseudo-items (from map/vehicle tiles, bionics etc) are considered
+         * @param limit stop searching after this many matches
+         * @note items must be empty to be considered a match
+         */
+        int amount_of( const std::string &what, bool pseudo = true, int limit = INT_MAX ) const;
+
+        /** Check instance provides at least qty of an item (@see amount_of) */
+        bool has_amount( const std::string &what, int qty, bool pseudo = true ) const {
+            return amount_of( what, pseudo, qty ) == qty;
```

Comment:
Not wrong, but I recall some old variant of `amount_of` understanding early quitting. Crafting inventories can get huge, so early quitting makes sense.